fusequery/query/src/servers/clickhouse/interactive_worker_base.rs

```diff
@@ -25,42 +33,106 @@ pub struct InteractiveWorkerBase;
+                let cancel = Arc::new(AtomicBool::new(false));
 
-        tokio::spawn(async move {
-            while !cancel.load(Ordering::Relaxed) {
-                let _ = interval_stream.next().await;
-                tx.send(BlockItem::ProgressTicker).await.ok();
-            }
-        });
+                let (mut tx, rx) = mpsc::channel(20);
+                let mut tx2 = tx.clone();
+                let cancel_clone = cancel.clone();
 
-        ctx.execute_task(async move {
-            while let Some(block) = abort_stream.next().await {
-                tx2.send(BlockItem::Block(block)).await.ok();
+                tokio::spawn(async move {
+                    while !cancel.load(Ordering::Relaxed) {
+                        let _ = interval_stream.next().await;
+                        tx.send(BlockItem::ProgressTicker).await.ok();
+                    }
+                });
+
+                ctx.execute_task(async move {
+                    while let Some(block) = abort_stream.next().await {
+                        tx2.send(BlockItem::Block(block)).await.ok();
+                    }
+
+                    cancel_clone.store(true, Ordering::Relaxed);
+                })?;
+
+                Ok(rx)
             }
+        }
+    }
 
-            cancel_clone.store(true, Ordering::Relaxed);
-        })?;
+    pub async fn process_insert_query(
+        insert: InsertIntoPlan,
+        ch_ctx: &mut CHContext,
+        ctx: FuseQueryContextRef,
+    ) -> Result<Receiver<BlockItem>> {
+        let sample_block = DataBlock::empty_with_schema(insert.schema());
+        let (sender, rec) = channel(4);
+        ch_ctx.state.out = Some(sender);
+
+        let sc = sample_block.schema().clone();
+        let stream = ReceiverStream::new(rec);
+        let stream = FromClickHouseBlockStream {
+            input: stream,
+            schema: sc,
+        };
+        insert.set_input_stream(Box::pin(stream));
+        let interpreter = InterpreterFactory::get(ctx.clone(), PlanNode::InsertInto(insert))?;
+
+        let (mut tx, rx) = mpsc::channel(20);
+        tx.send(BlockItem::InsertSample(sample_block)).await.ok();
 
+        // the data is comming in async mode
+        let sent_all_data = ch_ctx.state.sent_all_data.clone();
+        ctx.execute_task(async move {
+            interpreter.execute().await.unwrap();
+            sent_all_data.notify_one();
+        })?;
         Ok(rx)
     }
 }
+
+pub struct FromClickHouseBlockStream {
+    input: ReceiverStream<ClickHouseBlock>,
+    schema: DataSchemaRef,
+}
+
+impl futures::stream::Stream for FromClickHouseBlockStream {
+    type Item = DataBlock;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        self.input.poll_next_unpin(cx).map(|x| match x {
+            Some(v) => {
+                let block = from_clickhouse_block(self.schema.clone(), v).unwrap();
```

Comment:
Hm, any way to handle the error instead of unwrap?

Reply:
Can add a `log:error!` like flight did in  https://github.com/datafuselabs/datafuse/blob/master/common/flights/src/impls/storage_api_impl.rs#L108-L114